src/pyroboplan/trajectory/trajectory_optimization.py

```diff
@@ -259,12 +304,122 @@ def _eval_jerk(self, x_d, xc_d, h, k, n, step):
 
         Return
         ------
-            pydrake.autodiffutils.AutodiffXd
+            pydrake.autodiffutils.AutoDiffXd
                 The jerk along the specified segment evaluated at the specified step.
         """
         return 8.0 * (x_d[k, n] - 2.0 * xc_d[k, n] + x_d[k + 1, n]) / h[k] ** 2
 
-    def plan(self, q_path):
+    def _collision_constraint(
+        self,
+        q_val,
+        num_waypoints,
+        num_dofs,
+        collision_pairs,
+        min_allowable_dist,
+        influence_dist,
+    ):
+        """
+        Helper function to evaluate collision constraint and its gradients.
+
+        These calculations are based off the following resources:
+          * https://typeset.io/pdf/a-collision-free-mpc-for-whole-body-dynamic-locomotion-and-1l6itpfk.pdf
+          * https://laas.hal.science/hal-04425002
+
+        Parameters
+        ----------
+            q_val : pydrake.autodiffutils.AutoDiffXd
+                The joint configuration at each waypoint.
+            num_waypoints : int
+                The number of waypoints.
+            num_dofs : int
+                The number of degrees of freedom.
+            collision_pairs : list[list[int]]
+                A nested list containing the indices of collision model pairs for each geometry to check.
+            min_allowable_dist : float
+                The minimum allowable collision distance, in meters, for the constraint to be satisfied.
+                Positive is out of collision and negative is in collision.
+            influence_dist : float
+                The distance over which the collision distance has an effect on the constraint gradient.
+                This value should be larger than `min_allowable_dist`.
+
+        Return
+        ------
+            pydrake.autodiffutils.AutoDiffXd
+                The constraint value representing the minimum collision distance for every collision object and waypoint,
+                as well as its corresponding gradient.
+        """
+        if q_val.dtype != float:
```

Comment:
If you're willing to break this up even more it might help with readability, even though I know it's already a helper function... There's a lot of tabs in here.

Also this will return None if this check fails? Is that what you want?

Reply:
Maybe I shouldn't even do the datatype check and assume it's only ever gonna be used with the AutoDiff types during optimization?